exec/cnex/httpserver.c

```diff
@@ -0,0 +1,353 @@
+    sin.sin_family = AF_INET;
+    sin.sin_addr.s_addr = INADDR_ANY;
+    sin.sin_port = htons(port);
+    r = bind(server->server, (struct sockaddr *)&sin, sizeof(sin));
+    if (r != 0) {
+        // The executor will continue to run; just without debugger support.
+        perror("Could not bind debug server port.  Debugger will not be available.");
+        closesocket(server->server);
+        server->server = INVALID_SOCKET;
+        free(server);
+        return NULL;
+    }
+    r = listen(server->server, 5);
+    assert(r == 0);
+    return server;
+}
+
+void server_removeClient(HttpServerImpl *self, size_t index)
+{
+    Client *c = self->clients[index];
+    for(size_t i = index; i+1 < self->num_clients; i++) {
+        self->clients[i] = self->clients[i+1];
+    }
+    self->num_clients--;
+    self->clients = realloc(self->clients, (sizeof(Client*) * self->num_clients));
+    client_freeClient(c);
+}
+
+void server_service(HttpServerImpl *impl, BOOL wait)
+{
+    server_impl(impl, wait);
+}
+
+void server_freeServer(HttpServerImpl *impl)
+{
+    if (impl == NULL) {
+        return;
+    }
+    closesocket(impl->server);
+    for (size_t i = 0; i < impl->num_clients; i++) {
+        client_freeClient(impl->clients[i]);
+    }
+    free(impl->clients);
+    free(impl);
+}
+
+
+void server_impl(HttpServerImpl *impl, BOOL wait)
+{
+    if (impl->server == INVALID_SOCKET) {
+        return;
+    }
+    fd_set rfds;
+    FD_ZERO(&rfds);
+    FD_SET(impl->server, &rfds);
+    SOCKET maxsocket = impl->server;
+    for (size_t i = 0; i < impl->num_clients; i++) {
+        FD_SET(impl->clients[i]->socket, &rfds);
+        if (impl->clients[i]->socket > maxsocket) {
+            maxsocket = impl->clients[i]->socket;
+        }
+    }
+    struct timeval tv = {0, 0};
+    int n = select((int)maxsocket+1, &rfds, NULL, NULL, wait ? NULL : &tv);
+    if (n > 0) {
+        if (FD_ISSET(impl->server, &rfds)) {
+            struct sockaddr_in sin;
+            socklen_t socklen = sizeof(sin);
+            SOCKET t = accept(impl->server, (struct sockaddr *)&sin, &socklen);
+            assert(t != INVALID_SOCKET);
+            Client *c = client_createClient(impl->exec, impl->handler, t);
+            if (impl->clients) {
+                impl->clients = realloc(impl->clients, sizeof(Client*) * (impl->num_clients + 1));
+                impl->num_clients++;
+            }
+            if (impl->clients == NULL) {
+                impl->clients = malloc(sizeof(Client*));
+                impl->num_clients = 1;
+            }
+            impl->clients[impl->num_clients-1] = c;
+        }
+        for (size_t i = 0; i < impl->num_clients; i++) {
+            TString *s = string_newString();
+            Client *c = impl->clients[i];
+            if (FD_ISSET(c->socket, &rfds)) {
+                char buf[200];
+                n = recv(c->socket, buf, sizeof(buf), 0);
+                if (n > 0) {
+                    s = string_appendCString(s, buf);
```

Comment:
Rather than allocating `s` up above before you know you need it, and having to free it each iteration through the loop even though it might never have been used, it would probably be better to allocate it here based on `buf` and `n` directly. Also you've got a potential buffer overread here because recv does not nul-terminate anything and you're assuming that `buf` is a C string (it's not).

Reply:
Yes.  I agree there!  Fixed this to use `string_createStringFromData()` to avoid any buffer issues.  I'm actually surprised that this never triggered any warnings from the Sanitizers.